[DSM] Add datastreams monitoring to Azure Service Bus integration

zacharycmontoya opened the pull request:
## Summary of changes
This builds on top of the tracing integration for `Azure.Messaging.ServiceBus v7.14.0` to propagate DSM information over Azure Service Bus messages. The implementation uses bytecode instrumentation to directly access the message objects before being published/consumed by the library.

## Reason for change
Azure Service Bus is a common use-case for our users that use APM and DSM.

## Implementation details
### On Send
The send instrumentation is convoluted because the Activity handler which provides all of the tracing information _does not_ provide access to the message objects or its properties (aka headers). As a result, there are three times inside a message sending operation where we intervene.

#### The Inner Workings
Starting from the deepest part of the callstack:
1. The Activity for each ServiceBusMessage is started and stopped entirely within the `MessagingClientDiagnostics.InstrumentMessage` method. The Activity does not have direct access to a `ServiceBusMessage` or its properties dictionary.
2. The `MessagingClientDiagnostics.InstrumentMessage` method accepts one IDictionary<string,object> object, which corresponds to the properties of one `ServiceBusMessage`. The method starts and stops a new Activity and then adds the trace context information to the input dictionary.
3. There are two callers of `MessagingClientDiagnostics.InstrumentMessage` which have access to the actual ServiceBusMessage objects as their method arguments. They are
     - ServiceBusSender.CreateDiagnosticScope (input is an `IReadOnlyCollection<ServiceBusMessage>` object)
     - ServiceBusMessageBatch.TryAddMessage (input is one `ServiceBusMessage` object)

#### Instrumenting In This Crazy World
Due to the above factors, the implementation does the following:
1. The callsites `ServiceBusSender.CreateDiagnosticScope` and `ServiceBusMessageBatch.TryAddMessage` will create a mapping in a `CondtionalWeakTable` between the message properties dictionary and its owning `ServiceBusMessage`. (Uses bytecode instrumentation)
2. `MessagingClientDiagnostics.InstrumentMessage` will place the dictionary into a ThreadStatic variable, since each message is individually instrumented using `Activity`. (Uses bytecode instrumentation)
3. When the `Activity` is stopped, our `AzureServiceBusActivityHandler` will grab the dictionary from the ThreadStatic variable, access the `ConditionalWeakTable` to get the corresponding `ServiceBusMessage` object, and then use the `ServiceBusMessage` to calculate the PayloadSize, calculate the pathway context, add it to the span, and inject the pathway context back into the message properties. (Uses ActivityHandler)

### On Consume
The implementation is entirely contained in `ProcessMessageIntegration.cs`. The incoming ServiceBusMessage object is passed as input to the method, so we can extract the pathway context from the message properties and apply them to the current Datadog span.

## Test coverage
This PR adds a `Samples.DataStreams.AzureServiceBus` test application and snapshot testing to validate the end-to-end datastreams propagation. Outside testing will be required to confirm that the chosen tags result in the proper UI experience.

zacharycmontoya commented on tracer/src/Datadog.Trace/Activity/Handlers/AzureServiceBusActivityHandler.cs:
```diff
@@ -24,11 +28,69 @@ public void ActivityStarted<T>(string sourceName, T activity)
+
+                    var edgeTags = string.IsNullOrEmpty(namespaceString)
+                        ? new[] { "direction:out", "type:azureservicebus" }
+                        : new[] { "direction:out", $"topic:{namespaceString}", "type:azureservicebus" };
```
For Azure Service Bus, you can either:
1. Send to a Queue and receive from that Queue
2. Send to a Topic (which acts like an exchange) and receive from a Subscription

Ideally, I would do the following:
- Queue: `topic:<queueName>`
- Topic: `exchange:<topicName>`

If I can't figure out a way to disambiguate the two, what should I do? When you send a message using the library, it treats your string input as just a `string EntityPath`, so I don't know if we can actually tell on the client side if the message is going to a queue or a topic. I'll keep digging to see if it's possible to determine but at the moment I'm not sure if we will be able to

zacharycmontoya replied:
See this docs link for more info: https://learn.microsoft.com/en-us/dotnet/api/azure.messaging.servicebus.servicebussender?view=azure-dotnet

zacharycmontoya commented on tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Azure/ServiceBus/ProcessMessageIntegration.cs:
```diff
@@ -0,0 +1,102 @@
+
+                    // TODO: we could pool these arrays to reduce allocations
+                    // NOTE: the tags must be sorted in alphabetical order
+                    var edgeTags = string.IsNullOrEmpty(namespaceString)
```
As I mentioned in this [other comment](https://github.com/DataDog/dd-trace-dotnet/pull/4576/files#r1312381878), an incoming message can come directly from a queue or from a subscription (to a topic). Do we need to have a different tag for each type? The good thing is the subscription has a structure like `<topicName>/Subscriptions/<subscriptionName>` so if we need to distinguish between receiving from a queue and a subscription, it can be done.

andrewlock reviewed (approved):
Wow, nice work! That one made my brain hurt 😂 LGTM, just small potential nit

kr-igor commented on tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Azure/ServiceBus/AzureServiceBusCommon.cs:
```diff
@@ -0,0 +1,69 @@
+            return size;
+        }
+
+        private static long TryGetValueSize(object obj)
```
nit: maybe there's a common place where methods like this can be? I believe this one is repeated at least 2 times in different integrations.

zacharycmontoya replied:
Added a shared DSM helper class with this method in 0637b21fcf3fad6d0adff76a89e83fed9b654e47

kr-igor reviewed (approved):
Thank you for working on this!

andrewlock commented:
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (4576)](https://github.com/DataDog/dd-trace-dotnet/tree/d6e37d2150a82a177acbe49b5032a138ea86f461)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4576) - mean (76ms)  : 63, 89
     .   : milestone, 76,
    master - mean (77ms)  : 57, 97
     .   : milestone, 77,

    section CallTarget+Inlining+NGEN
    This PR (4576) - mean (994ms)  : 968, 1020
     .   : milestone, 994,
    master - mean (996ms)  : 949, 1042
     .   : milestone, 996,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4576) - mean (105ms)  : 101, 109
     .   : milestone, 105,
    master - mean (104ms)  : 101, 108
     .   : milestone, 104,

    section CallTarget+Inlining+NGEN
    This PR (4576) - mean (687ms)  : 668, 706
     .   : milestone, 687,
    master - mean (691ms)  : 662, 721
     .   : milestone, 691,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4576) - mean (89ms)  : 86, 92
     .   : milestone, 89,
    master - mean (90ms)  : 82, 97
     .   : milestone, 90,

    section CallTarget+Inlining+NGEN
    This PR (4576) - mean (654ms)  : 624, 684
     .   : milestone, 654,
    master - mean (657ms)  : 625, 689
     .   : milestone, 657,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4576) - mean (190ms)  : 185, 196
     .   : milestone, 190,
    master - mean (190ms)  : 186, 194
     .   : milestone, 190,

    section CallTarget+Inlining+NGEN
    This PR (4576) - mean (1,089ms)  : 1059, 1118
     .   : milestone, 1089,
    master - mean (1,092ms)  : 1053, 1131
     .   : milestone, 1092,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4576) - mean (272ms)  : 264, 280
     .   : milestone, 272,
    master - mean (270ms)  : 258, 281
     .   : milestone, 270,

    section CallTarget+Inlining+NGEN
    This PR (4576) - mean (1,057ms)  : 1015, 1100
     .   : milestone, 1057,
    master - mean (1,047ms)  : 1013, 1081
     .   : milestone, 1047,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4576) - mean (261ms)  : 254, 268
     .   : milestone, 261,
    master - mean (261ms)  : 255, 267
     .   : milestone, 261,

    section CallTarget+Inlining+NGEN
    This PR (4576) - mean (1,003ms)  : 962, 1043
     .   : milestone, 1003,
    master - mean (1,003ms)  : 971, 1035
     .   : milestone, 1003,

```

andrewlock commented:
## Benchmarks Report :snail:

Benchmarks for #4576 compared to [master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c): 
* 2 benchmarks are **slower**, with geometric mean 1.257
* All benchmarks have **the same** allocations


The following thresholds were used for comparing the benchmark speeds:
* Mann–Whitney U test with statistical test for significance of **5%**
* Only results indicating a difference greater than **10%** and **0.3 ns** are considered.

Allocation changes below **0.5%** are ignored.


### Benchmark details


<details><summary><strong>Benchmarks.Trace.ActivityBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`StartStopWithChild`|net6.0|8.26μs|46.1ns|313ns|0.0325| 0.0163|0|7.3&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`StartStopWithChild`|netcoreapp3.1|10μs|54.1ns|306ns|0.0251| 0.01|0|7.4&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`StartStopWithChild`|net472|15.8μs|61.2ns|237ns|1.28| 0.315|0.102|7.68&nbsp;KB|
|#4576|`StartStopWithChild`|net6.0|8.28μs|40ns|160ns|0.02| 0.00802|0|7.29&nbsp;KB|
|#4576|`StartStopWithChild`|netcoreapp3.1|10.2μs|52.9ns|318ns|0.0256| 0.0102|0|7.39&nbsp;KB|
|#4576|`StartStopWithChild`|net472|15.7μs|54.3ns|210ns|1.28| 0.306|0.0942|7.66&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.AgentWriterBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`WriteAndFlushEnrichedTraces`|net6.0|490μs|351ns|1.36μs|0| 0|0|2.7&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|633μs|191ns|716ns|0| 0|0|2.7&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`WriteAndFlushEnrichedTraces`|net472|793μs|342ns|1.33μs|0.396| 0|0|3.3&nbsp;KB|
|#4576|`WriteAndFlushEnrichedTraces`|net6.0|467μs|280ns|1.08μs|0| 0|0|2.7&nbsp;KB|
|#4576|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|631μs|192ns|742ns|0| 0|0|2.7&nbsp;KB|
|#4576|`WriteAndFlushEnrichedTraces`|net472|790μs|369ns|1.38μs|0.398| 0|0|3.3&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Asm.AppSecBodyBenchmark</strong> - Slower :warning: Same allocations :heavy_check_mark:



#### Slower :warning: in #4576

|  Benchmark         | diff/base | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.Asm.AppSecBodyBenchmark.ObjectExtractorSimpleBody&#8209;net6.0 | 1.412 | 123.09 | 173.78 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`AllCycleSimpleBody`|net6.0|38.8μs|15.2ns|58.8ns|0.0195| 0|0|1.69&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`AllCycleSimpleBody`|netcoreapp3.1|41.6μs|50.2ns|181ns|0.0206| 0|0|1.67&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`AllCycleSimpleBody`|net472|43.3μs|26.9ns|101ns|0.259| 0|0|1.73&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`AllCycleMoreComplexBody`|net6.0|233μs|89.7ns|391ns|0.116| 0|0|9.26&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`AllCycleMoreComplexBody`|netcoreapp3.1|242μs|330ns|1.28μs|0| 0|0|9.16&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`AllCycleMoreComplexBody`|net472|250μs|85.8ns|321ns|1.37| 0|0|9.33&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`ObjectExtractorSimpleBody`|net6.0|123ns|0.051ns|0.198ns|0.00393| 0|0|280&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`ObjectExtractorSimpleBody`|netcoreapp3.1|170ns|0.353ns|1.37ns|0.00379| 0|0|272&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`ObjectExtractorSimpleBody`|net472|146ns|0.117ns|0.453ns|0.0446| 0|0|281&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`ObjectExtractorMoreComplexBody`|net6.0|3.07μs|1.29ns|4.64ns|0.0551| 0|0|3.88&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`ObjectExtractorMoreComplexBody`|netcoreapp3.1|3.99μs|1.22ns|4.55ns|0.0518| 0|0|3.78&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`ObjectExtractorMoreComplexBody`|net472|4.11μs|4.49ns|17.4ns|0.618| 0.00614|0|3.89&nbsp;KB|
|#4576|`AllCycleSimpleBody`|net6.0|38.9μs|18.5ns|71.7ns|0.0195| 0|0|1.69&nbsp;KB|
|#4576|`AllCycleSimpleBody`|netcoreapp3.1|40.8μs|98.5ns|394ns|0.02| 0|0|1.67&nbsp;KB|
|#4576|`AllCycleSimpleBody`|net472|43.6μs|9.13ns|32.9ns|0.258| 0|0|1.73&nbsp;KB|
|#4576|`AllCycleMoreComplexBody`|net6.0|230μs|60.2ns|225ns|0.115| 0|0|9.26&nbsp;KB|
|#4576|`AllCycleMoreComplexBody`|netcoreapp3.1|243μs|557ns|2.08μs|0| 0|0|9.16&nbsp;KB|
|#4576|`AllCycleMoreComplexBody`|net472|248μs|93.7ns|363ns|1.36| 0|0|9.33&nbsp;KB|
|#4576|`ObjectExtractorSimpleBody`|net6.0|174ns|0.0895ns|0.346ns|0.0039| 0|0|280&nbsp;B|
|#4576|`ObjectExtractorSimpleBody`|netcoreapp3.1|176ns|0.0617ns|0.231ns|0.00372| 0|0|272&nbsp;B|
|#4576|`ObjectExtractorSimpleBody`|net472|143ns|0.145ns|0.561ns|0.0446| 0|0|281&nbsp;B|
|#4576|`ObjectExtractorMoreComplexBody`|net6.0|2.99μs|0.971ns|3.76ns|0.0539| 0|0|3.88&nbsp;KB|
|#4576|`ObjectExtractorMoreComplexBody`|netcoreapp3.1|4.34μs|1.57ns|6.09ns|0.05| 0|0|3.78&nbsp;KB|
|#4576|`ObjectExtractorMoreComplexBody`|net472|4.11μs|3.98ns|15.4ns|0.618| 0.00616|0|3.89&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Asm.AppSecWafBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWaf(args=NestedMap (10))`|net6.0|83.2μs|47ns|182ns|0.217| 0|0|17.18&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWaf(args=NestedMap (10))`|netcoreapp3.1|98μs|508ns|2.59μs|0.228| 0|0|17.18&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWaf(args=NestedMap (10))`|net472|114μs|43ns|161ns|2.79| 0.114|0|17.82&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWafWithAttack(args=Neste(...)tack) [22])`|net6.0|121μs|512ns|1.92μs|0.348| 0|0|24.7&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWafWithAttack(args=Neste(...)tack) [22])`|netcoreapp3.1|130μs|202ns|728ns|0.324| 0|0|24.66&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWafWithAttack(args=Neste(...)tack) [22])`|net472|158μs|179ns|672ns|4.02| 0.158|0|25.56&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWaf(args=NestedMap (100))`|net6.0|161μs|105ns|394ns|0.52| 0|0|36.94&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWaf(args=NestedMap (100))`|netcoreapp3.1|188μs|567ns|2.04μs|0.484| 0|0|37.6&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWaf(args=NestedMap (100))`|net472|233μs|78ns|292ns|6.2| 0.46|0|39.5&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWafWithAttack(args=Neste(...)tack) [23])`|net6.0|198μs|83.5ns|312ns|0.532| 0|0|44.47&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWafWithAttack(args=Neste(...)tack) [23])`|netcoreapp3.1|237μs|1.22μs|5.44μs|0.488| 0|0|45.07&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWafWithAttack(args=Neste(...)tack) [23])`|net472|289μs|432ns|1.56μs|7.49| 0.576|0|47.25&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWaf(args=NestedMap (1000))`|net6.0|171μs|89.5ns|335ns|0.514| 0|0|36.94&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWaf(args=NestedMap (1000))`|netcoreapp3.1|187μs|603ns|2.63μs|0.464| 0|0|37.6&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWaf(args=NestedMap (1000))`|net472|229μs|176ns|659ns|6.26| 0.455|0|39.51&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWafWithAttack(args=Neste(...)tack) [24])`|net6.0|201μs|169ns|653ns|0.602| 0|0|44.46&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWafWithAttack(args=Neste(...)tack) [24])`|netcoreapp3.1|240μs|473ns|1.77μs|0.596| 0|0|45.07&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunWafWithAttack(args=Neste(...)tack) [24])`|net472|290μs|1.19μs|4.61μs|7.44| 0.572|0|47.24&nbsp;KB|
|#4576|`RunWaf(args=NestedMap (10))`|net6.0|85.1μs|331ns|1.24μs|0.207| 0|0|17.18&nbsp;KB|
|#4576|`RunWaf(args=NestedMap (10))`|netcoreapp3.1|95.3μs|463ns|1.96μs|0.23| 0|0|17.18&nbsp;KB|
|#4576|`RunWaf(args=NestedMap (10))`|net472|119μs|361ns|1.4μs|2.78| 0.114|0|17.82&nbsp;KB|
|#4576|`RunWafWithAttack(args=Neste(...)tack) [22])`|net6.0|116μs|83.7ns|302ns|0.348| 0|0|24.7&nbsp;KB|
|#4576|`RunWafWithAttack(args=Neste(...)tack) [22])`|netcoreapp3.1|132μs|616ns|2.46μs|0.325| 0|0|24.66&nbsp;KB|
|#4576|`RunWafWithAttack(args=Neste(...)tack) [22])`|net472|159μs|390ns|1.51μs|3.99| 0.156|0|25.56&nbsp;KB|
|#4576|`RunWaf(args=NestedMap (100))`|net6.0|168μs|125ns|486ns|0.474| 0|0|36.94&nbsp;KB|
|#4576|`RunWaf(args=NestedMap (100))`|netcoreapp3.1|196μs|1μs|4.91μs|0.481| 0|0|37.6&nbsp;KB|
|#4576|`RunWaf(args=NestedMap (100))`|net472|237μs|326ns|1.22μs|6.2| 0.468|0|39.5&nbsp;KB|
|#4576|`RunWafWithAttack(args=Neste(...)tack) [23])`|net6.0|204μs|937ns|3.63μs|0.594| 0|0|44.47&nbsp;KB|
|#4576|`RunWafWithAttack(args=Neste(...)tack) [23])`|netcoreapp3.1|244μs|1.21μs|5.43μs|0.608| 0|0|45.08&nbsp;KB|
|#4576|`RunWafWithAttack(args=Neste(...)tack) [23])`|net472|291μs|88ns|329ns|7.43| 0.583|0|47.24&nbsp;KB|
|#4576|`RunWaf(args=NestedMap (1000))`|net6.0|172μs|134ns|484ns|0.515| 0|0|36.94&nbsp;KB|
|#4576|`RunWaf(args=NestedMap (1000))`|netcoreapp3.1|197μs|699ns|2.71μs|0.484| 0|0|37.6&nbsp;KB|
|#4576|`RunWaf(args=NestedMap (1000))`|net472|241μs|533ns|2.06μs|6.19| 0.459|0|39.51&nbsp;KB|
|#4576|`RunWafWithAttack(args=Neste(...)tack) [24])`|net6.0|209μs|263ns|1.02μs|0.591| 0|0|44.46&nbsp;KB|
|#4576|`RunWafWithAttack(args=Neste(...)tack) [24])`|netcoreapp3.1|242μs|585ns|2.27μs|0.583| 0|0|45.07&nbsp;KB|
|#4576|`RunWafWithAttack(args=Neste(...)tack) [24])`|net472|293μs|143ns|552ns|7.4| 0.58|0|47.25&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.AspNetCoreBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`SendRequest`|net6.0|171μs|123ns|462ns|0.17| 0|0|18.09&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`SendRequest`|netcoreapp3.1|189μs|163ns|610ns|0.189| 0|0|20.25&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`SendRequest`|net472|0.000673ns|0.000228ns|0.000852ns|0| 0|0|0&nbsp;b|
|#4576|`SendRequest`|net6.0|168μs|76.5ns|286ns|0.169| 0|0|18.09&nbsp;KB|
|#4576|`SendRequest`|netcoreapp3.1|191μs|258ns|998ns|0.189| 0|0|20.25&nbsp;KB|
|#4576|`SendRequest`|net472|0.0026ns|0.0011ns|0.00412ns|0| 0|0|0&nbsp;b|
</details>


<details><summary><strong>Benchmarks.Trace.CIVisibilityProtocolWriterBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`WriteAndFlushEnrichedTraces`|net6.0|515μs|1.13μs|4.39μs|0.498| 0|0|41.38&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|628μs|1.32μs|5.11μs|0.308| 0|0|41.58&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`WriteAndFlushEnrichedTraces`|net472|789μs|2.85μs|11μs|8.38| 2.67|0.381|53.23&nbsp;KB|
|#4576|`WriteAndFlushEnrichedTraces`|net6.0|510μs|2.21μs|8.56μs|0.53| 0|0|41.51&nbsp;KB|
|#4576|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|617μs|1.73μs|6.72μs|0.308| 0|0|41.71&nbsp;KB|
|#4576|`WriteAndFlushEnrichedTraces`|net472|768μs|2.32μs|9μs|8.33| 2.65|0.379|53.22&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.DbCommandBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`ExecuteNonQuery`|net6.0|1.01μs|0.608ns|2.35ns|0.0107| 0|0|768&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`ExecuteNonQuery`|netcoreapp3.1|1.34μs|0.531ns|2.06ns|0.0101| 0|0|768&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`ExecuteNonQuery`|net472|1.6μs|0.812ns|3.04ns|0.115| 0|0|730&nbsp;B|
|#4576|`ExecuteNonQuery`|net6.0|1.07μs|0.521ns|2.02ns|0.0107| 0|0|768&nbsp;B|
|#4576|`ExecuteNonQuery`|netcoreapp3.1|1.38μs|0.587ns|2.2ns|0.0104| 0|0|768&nbsp;B|
|#4576|`ExecuteNonQuery`|net472|1.65μs|2.04ns|7.88ns|0.116| 0|0|730&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.ElasticsearchBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`CallElasticsearch`|net6.0|1.19μs|1.32ns|4.92ns|0.0137| 0|0|992&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`CallElasticsearch`|netcoreapp3.1|1.54μs|0.458ns|1.59ns|0.013| 0|0|992&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`CallElasticsearch`|net472|2.4μs|0.696ns|2.51ns|0.159| 0|0|1&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`CallElasticsearchAsync`|net6.0|1.28μs|0.3ns|1.04ns|0.0136| 0|0|968&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`CallElasticsearchAsync`|netcoreapp3.1|1.58μs|0.83ns|3.21ns|0.0143| 0|0|1.04&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`CallElasticsearchAsync`|net472|2.58μs|0.918ns|3.56ns|0.168| 0.0013|0|1.06&nbsp;KB|
|#4576|`CallElasticsearch`|net6.0|1.26μs|1.51ns|5.86ns|0.0137| 0|0|992&nbsp;B|
|#4576|`CallElasticsearch`|netcoreapp3.1|1.4μs|0.407ns|1.47ns|0.0132| 0|0|992&nbsp;B|
|#4576|`CallElasticsearch`|net472|2.35μs|1.12ns|4.21ns|0.158| 0|0|1&nbsp;KB|
|#4576|`CallElasticsearchAsync`|net6.0|1.39μs|0.497ns|1.86ns|0.0133| 0|0|968&nbsp;B|
|#4576|`CallElasticsearchAsync`|netcoreapp3.1|1.5μs|0.545ns|2.04ns|0.0143| 0|0|1.04&nbsp;KB|
|#4576|`CallElasticsearchAsync`|net472|2.54μs|0.617ns|2.31ns|0.168| 0|0|1.06&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.GraphQLBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`ExecuteAsync`|net6.0|1.19μs|0.772ns|2.99ns|0.0126| 0|0|912&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`ExecuteAsync`|netcoreapp3.1|1.43μs|0.406ns|1.52ns|0.0121| 0|0|912&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`ExecuteAsync`|net472|1.71μs|0.574ns|2.15ns|0.139| 0.000847|0|875&nbsp;B|
|#4576|`ExecuteAsync`|net6.0|1.28μs|0.564ns|2.11ns|0.0128| 0|0|912&nbsp;B|
|#4576|`ExecuteAsync`|netcoreapp3.1|1.51μs|0.727ns|2.82ns|0.0119| 0|0|912&nbsp;B|
|#4576|`ExecuteAsync`|net472|1.58μs|1.35ns|5.21ns|0.138| 0|0|875&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.HttpClientBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`SendAsync`|net6.0|3.84μs|1.05ns|3.93ns|0.0269| 0|0|1.94&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`SendAsync`|netcoreapp3.1|4.54μs|2.04ns|7.64ns|0.0339| 0|0|2.48&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`SendAsync`|net472|7.14μs|2.63ns|10.2ns|0.481| 0|0|3.05&nbsp;KB|
|#4576|`SendAsync`|net6.0|3.8μs|1.4ns|5.44ns|0.0269| 0|0|1.94&nbsp;KB|
|#4576|`SendAsync`|netcoreapp3.1|4.34μs|2.34ns|9.05ns|0.0326| 0|0|2.48&nbsp;KB|
|#4576|`SendAsync`|net472|7.14μs|3.79ns|14.7ns|0.484| 0|0|3.05&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.ILoggerBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`EnrichedLog`|net6.0|1.37μs|0.74ns|2.87ns|0.0227| 0|0|1.62&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`EnrichedLog`|netcoreapp3.1|1.89μs|0.588ns|2.28ns|0.0219| 0|0|1.62&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`EnrichedLog`|net472|2.21μs|2.51ns|9.73ns|0.244| 0|0|1.54&nbsp;KB|
|#4576|`EnrichedLog`|net6.0|1.4μs|0.99ns|3.83ns|0.0228| 0|0|1.62&nbsp;KB|
|#4576|`EnrichedLog`|netcoreapp3.1|1.85μs|0.87ns|3.26ns|0.0222| 0|0|1.62&nbsp;KB|
|#4576|`EnrichedLog`|net472|2.43μs|2.42ns|9.39ns|0.244| 0|0|1.54&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Log4netBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`EnrichedLog`|net6.0|112μs|71ns|265ns|0| 0|0|4.21&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`EnrichedLog`|netcoreapp3.1|117μs|142ns|551ns|0.0585| 0|0|4.21&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`EnrichedLog`|net472|148μs|106ns|410ns|0.663| 0.221|0|4.38&nbsp;KB|
|#4576|`EnrichedLog`|net6.0|112μs|81.4ns|315ns|0.0564| 0|0|4.21&nbsp;KB|
|#4576|`EnrichedLog`|netcoreapp3.1|119μs|199ns|771ns|0.0588| 0|0|4.21&nbsp;KB|
|#4576|`EnrichedLog`|net472|149μs|108ns|404ns|0.666| 0.222|0|4.38&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.NLogBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`EnrichedLog`|net6.0|2.96μs|0.697ns|2.61ns|0.0296| 0|0|2.18&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`EnrichedLog`|netcoreapp3.1|3.88μs|1.13ns|4.07ns|0.0293| 0|0|2.18&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`EnrichedLog`|net472|4.58μs|9.17ns|35.5ns|0.316| 0|0|1.99&nbsp;KB|
|#4576|`EnrichedLog`|net6.0|2.96μs|0.987ns|3.69ns|0.0297| 0|0|2.18&nbsp;KB|
|#4576|`EnrichedLog`|netcoreapp3.1|3.96μs|1.23ns|4.61ns|0.0295| 0|0|2.18&nbsp;KB|
|#4576|`EnrichedLog`|net472|4.5μs|1.96ns|7.59ns|0.315| 0|0|1.99&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.RedisBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`SendReceive`|net6.0|1.35μs|1.17ns|4.53ns|0.0163| 0|0|1.16&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`SendReceive`|netcoreapp3.1|1.72μs|0.89ns|3.33ns|0.0163| 0|0|1.16&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`SendReceive`|net472|2.07μs|2.03ns|7.61ns|0.184| 0|0|1.16&nbsp;KB|
|#4576|`SendReceive`|net6.0|1.37μs|1.26ns|4.71ns|0.0164| 0|0|1.16&nbsp;KB|
|#4576|`SendReceive`|netcoreapp3.1|1.6μs|1.41ns|5.08ns|0.0159| 0|0|1.16&nbsp;KB|
|#4576|`SendReceive`|net472|2.04μs|1.12ns|4.33ns|0.184| 0|0|1.16&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.SerilogBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`EnrichedLog`|net6.0|2.61μs|0.805ns|3.01ns|0.0209| 0|0|1.53&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`EnrichedLog`|netcoreapp3.1|3.61μs|0.956ns|3.45ns|0.0215| 0|0|1.58&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`EnrichedLog`|net472|3.99μs|2.05ns|7.69ns|0.309| 0|0|1.96&nbsp;KB|
|#4576|`EnrichedLog`|net6.0|2.58μs|0.522ns|1.95ns|0.0219| 0|0|1.53&nbsp;KB|
|#4576|`EnrichedLog`|netcoreapp3.1|3.41μs|1.96ns|7.32ns|0.0205| 0|0|1.58&nbsp;KB|
|#4576|`EnrichedLog`|net472|3.98μs|1.9ns|7.13ns|0.311| 0|0|1.96&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.SpanBenchmark</strong> - Slower :warning: Same allocations :heavy_check_mark:



#### Slower :warning: in #4576

|  Benchmark         | diff/base | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.SpanBenchmark.StartFinishSpan&#8209;net6.0 | 1.118 | 383.85 | 429.27 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`StartFinishSpan`|net6.0|384ns|0.187ns|0.701ns|0.00746| 0|0|536&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`StartFinishSpan`|netcoreapp3.1|545ns|0.218ns|0.845ns|0.00727| 0|0|536&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`StartFinishSpan`|net472|614ns|0.157ns|0.587ns|0.0854| 0|0|538&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`StartFinishScope`|net6.0|501ns|0.105ns|0.379ns|0.00907| 0|0|656&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`StartFinishScope`|netcoreapp3.1|703ns|0.938ns|3.63ns|0.009| 0|0|656&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`StartFinishScope`|net472|892ns|0.677ns|2.62ns|0.098| 0|0|618&nbsp;B|
|#4576|`StartFinishSpan`|net6.0|429ns|0.142ns|0.531ns|0.00756| 0|0|536&nbsp;B|
|#4576|`StartFinishSpan`|netcoreapp3.1|562ns|0.431ns|1.61ns|0.00733| 0|0|536&nbsp;B|
|#4576|`StartFinishSpan`|net472|624ns|0.233ns|0.904ns|0.0852| 0|0|538&nbsp;B|
|#4576|`StartFinishScope`|net6.0|543ns|0.202ns|0.781ns|0.0093| 0|0|656&nbsp;B|
|#4576|`StartFinishScope`|netcoreapp3.1|732ns|0.292ns|1.09ns|0.00858| 0|0|656&nbsp;B|
|#4576|`StartFinishScope`|net472|844ns|0.481ns|1.86ns|0.098| 0|0|618&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.TraceAnnotationsBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunOnMethodBegin`|net6.0|537ns|0.566ns|2.19ns|0.00914| 0|0|656&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunOnMethodBegin`|netcoreapp3.1|783ns|1.57ns|5.86ns|0.00899| 0|0|656&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)|`RunOnMethodBegin`|net472|956ns|0.486ns|1.88ns|0.0979| 0|0|618&nbsp;B|
|#4576|`RunOnMethodBegin`|net6.0|566ns|0.177ns|0.684ns|0.00936| 0|0|656&nbsp;B|
|#4576|`RunOnMethodBegin`|netcoreapp3.1|851ns|0.683ns|2.56ns|0.0091| 0|0|656&nbsp;B|
|#4576|`RunOnMethodBegin`|net472|912ns|0.387ns|1.45ns|0.0979| 0|0|618&nbsp;B|
</details>

andrewlock commented:
## Throughput/Crank Report:zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (4576)](https://github.com/DataDog/dd-trace-dotnet/tree/d6e37d2150a82a177acbe49b5032a138ea86f461)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)
- [benchmarks/2.37.0](https://github.com/DataDog/dd-trace-dotnet/tree/1cbf0f91defaa3fa262540d88bee223bc4a86732)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4576) (11.378M)   : 0, 11378421
    master (11.239M)   : 0, 11239346
    benchmarks/2.37.0 (10.984M)   : 0, 10983597
    benchmarks/2.9.0 (11.237M)   : 0, 11237158

    section Automatic
    This PR (4576) (7.882M)   : 0, 7881729
    master (7.864M)   : 0, 7863833
    benchmarks/2.37.0 (7.620M)   : 0, 7620446
    benchmarks/2.9.0 (7.980M)   : 0, 7979544

    section Trace stats
    master (7.823M)   : 0, 7822764
    benchmarks/2.37.0 (7.601M)   : 0, 7601244

    section Manual
    This PR (4576) (9.915M)   : 0, 9914710
    master (10.001M)   : 0, 10000648
    benchmarks/2.37.0 (9.672M)   : 0, 9672017

    section Manual + Automatic
    This PR (4576) (7.607M)   : 0, 7606679
    master (7.548M)   : 0, 7547695
    benchmarks/2.37.0 (7.271M)   : 0, 7270501

    section Version Conflict
    master (6.723M)   : 0, 6723060
    benchmarks/2.37.0 (6.624M)   : 0, 6624248

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4576) (9.623M)   : 0, 9622817
    master (9.679M)   : 0, 9678607
    benchmarks/2.37.0 (9.603M)   : 0, 9603393
    benchmarks/2.9.0 (9.575M)   : 0, 9574635

    section Automatic
    This PR (4576) (6.712M)   : 0, 6711656
    master (6.864M)   : 0, 6864284
    benchmarks/2.37.0 (6.484M)   : 0, 6483840

    section Trace stats
    master (6.863M)   : 0, 6863294
    benchmarks/2.37.0 (6.737M)   : 0, 6736780

    section Manual
    This PR (4576) (8.255M)   : crit ,0, 8255489
    master (8.743M)   : 0, 8743073
    benchmarks/2.37.0 (8.509M)   : 0, 8508994

    section Manual + Automatic
    This PR (4576) (6.565M)   : 0, 6565087
    master (6.636M)   : 0, 6636247
    benchmarks/2.37.0 (6.542M)   : 0, 6542042

    section Version Conflict
    master (5.812M)   : 0, 5811903
    benchmarks/2.37.0 (5.769M)   : 0, 5769174

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4576) (9.171M)   : 0, 9171262
    master (9.307M)   : 0, 9306588
    benchmarks/2.37.0 (9.641M)   : 0, 9641281
    benchmarks/2.9.0 (9.346M)   : 0, 9345561

    section Automatic
    This PR (4576) (6.667M)   : 0, 6667126
    master (4.429M)   : 0, 4429469
    benchmarks/2.37.0 (6.809M)   : 0, 6809282
    benchmarks/2.9.0 (6.752M)   : 0, 6751930

    section Trace stats
    master (6.582M)   : 0, 6582183
    benchmarks/2.37.0 (6.650M)   : 0, 6649920

    section Manual
    This PR (4576) (8.330M)   : 0, 8329806
    master (8.260M)   : 0, 8259686
    benchmarks/2.37.0 (8.352M)   : 0, 8352046

    section Manual + Automatic
    This PR (4576) (6.416M)   : 0, 6415810
    master (6.462M)   : 0, 6462403
    benchmarks/2.37.0 (6.611M)   : 0, 6611182

    section Version Conflict
    master (5.973M)   : 0, 5973007
    benchmarks/2.37.0 (6.067M)   : 0, 6067045

```
```mermaid
gantt
    title Throughput Linux x64 (ASM) (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4576) (7.357M)   : 0, 7357154
    master (7.204M)   : 0, 7204398
    benchmarks/2.37.0 (7.626M)   : 0, 7626233
    benchmarks/2.9.0 (7.928M)   : 0, 7928406

    section No attack
    This PR (4576) (2.128M)   : 0, 2127708
    master (2.090M)   : 0, 2090301
    benchmarks/2.37.0 (2.192M)   : 0, 2192041
    benchmarks/2.9.0 (3.365M)   : 0, 3365217

    section Attack
    This PR (4576) (1.691M)   : 0, 1691096
    master (1.647M)   : 0, 1646927
    benchmarks/2.37.0 (1.697M)   : 0, 1696915
    benchmarks/2.9.0 (2.632M)   : 0, 2631585

    section Blocking
    This PR (4576) (3.122M)   : 0, 3121500
    master (3.015M)   : 0, 3014683
    benchmarks/2.37.0 (3.146M)   : 0, 3146410

```

zacharycmontoya commented on Datadog.Trace.sln:
Fixed in the other PR 👍🏼

zacharycmontoya commented on tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsContextPropagator.cs:
```diff
@@ -47,4 +49,46 @@ public void Inject<TCarrier>(PathwayContext context, TCarrier headers)
+    {
+        if (headers is null) { ThrowHelper.ThrowArgumentNullException(nameof(headers)); }
+
+        headers.Add(DataStreamsPropagationHeaders.PropagationKey, Convert.ToBase64String(PathwayContextEncoder.Encode(context)));
```
The main limitation I had when trying to set the pathway context in the message headers is that I can't directly pass in a byte array, so we have to pass in a string of some sort. So the question becomes what encoding should we use. base64 made sense to me since we can easily encode/decode the 8-bit values this way, but I'm following up with the DSM team to confirm.

zacharycmontoya commented on tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Azure/ServiceBus/ServiceBusHeadersCollectionAdapter.cs:
```diff
@@ -0,0 +1,50 @@
+
+internal readonly struct ServiceBusHeadersCollectionAdapter : IHeadersCollection
+{
+    private static readonly IDatadogLogger Logger = DatadogLogging.GetLoggerFor<ServiceBusHeadersCollectionAdapter>();
```
Thanks! Fixed in 7fc692fb6b4a1614d3b09cb630c0a96c9965e826

zacharycmontoya commented on tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Azure/ServiceBus/AzureServiceBusCommon.cs:
```diff
@@ -0,0 +1,69 @@
+        [ThreadStatic]
+#pragma warning disable SA1401 // Fields should be private
+        internal static IDictionary<string, object> ActiveMessageProperties;
+#pragma warning restore SA1401 // Fields should be private
```
OMG I swear I tested it before and it was not working. Alas, it actually does work so I've made the change in bc4e7b23ad198af45b65cd8262fc6f21ac301ce6 . Thanks for pressing on this 😄

zacharycmontoya commented on tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Azure/ServiceBus/ProcessMessageIntegration.cs:
```diff
@@ -0,0 +1,102 @@
+                var tracer = Tracer.Instance;
+                var dataStreamsManager = tracer.TracerManager.DataStreamsManager;
+
+                if (dataStreamsManager.IsEnabled)
```
Refactoring done in f35efa466776ad6a6f3840db7673d25da4a451ff

zacharycmontoya commented on tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Azure/ServiceBus/SendServiceBusMessagesIntegration.cs:
```diff
@@ -0,0 +1,66 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System.Collections;
```
Ok I think I got them all in a8b37caf6fa278d81b30e137c48f5666d737406e

zacharycmontoya commented on tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/Azure/DataStreamsMonitoringAzureServiceBusTests.cs:
```diff
@@ -0,0 +1,154 @@
+                    _.MemberConverter<MockDataStreamsStatsPoint, byte[]>(
+                        x => x.PayloadSize,
+                        (_, v) =>  v?.Length == 0 ? v : new byte[] { 0xFF });
+                });
```
Done in 405e3328d598a9100dca6968de31ae73f9054a37

zacharycmontoya commented:
After rebasing and addressing feedback, the only remaining feedback left that may change is the string encoding of the pathway context byte array when it is passed as a message property. Since we cannot directly pass a byte array, we need to figure out what encoding to use. Currently it is base64, but if it changes due to our internal conversations, I'll update the encoding.

andrewlock commented on tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Azure/ServiceBus/IBody.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Azure.ServiceBus;
+
+internal interface IBody
```
> so I think this chain of interfaces is optimal.

FWIW, I don't think the duck chaining works like that (@tonyredondo can correct me). AFAIK, there's no benefit to making a proxy that is _only_ duck-chained (not used directly in a generic constraint) an interface vs a struct (besides the usual constraints of "if you need to invoke methods, needs to be an interface etc).

That said, it's going to be marginal benefits, so not a big deal at all 🙂 Also, I've recently learned there's differences in how nulls are handled in duck chaining depending on whether you use an interface or a struct, so there's that to contend with too 😅
